### DISCUSSION TITLE
Weird loop optimizations with C++

### DISCUSSION THREAD

**Discussion by DelinWorks** [axmolengine/axmol#643] [General]
One question about cpp is when I was working on improving the particle system, there were comments saying that separating loops to fill each data field was the best optimization, like:
![image](https://user-images.githubusercontent.com/45469625/170127949-5de77be5-4f3d-4f79-82dd-33796af7597d.png)

And then I was told that `std::fill_n` is even better for static values, like:
![image](https://user-images.githubusercontent.com/45469625/170128021-8fdc90d0-4c5c-46df-8cfb-9a7b07b6599e.png)

Soo I've ran a test and the results were shocking: 

![image](https://user-images.githubusercontent.com/45469625/170128121-8d023133-a14d-4bf2-9c53-8d47851a1115.png)

Why is the method of combining the loops into a single one the fastest? even though people say that separating loops is faster because cache hit rates are better.

A single loop is something like this:
![image](https://user-images.githubusercontent.com/45469625/170128412-9cf2eacb-8562-4039-b4b4-bc726ffedb13.png)

Can someone explain to me why is this the case?

The code:
```
#include <iostream>
#include "debug.hpp"

class bigData {
public:


	float* dummy1;
	float* dummy2;
	float* dummy3;

	bigData(int count);
};

int main() {

	auto watch = StopWatch();

	int count = 10000000;

	watch.start();
	auto d = bigData(count);
	watch.stop();

	printf("Allocating data took: %dms\n", watch.getElapsed());

	watch.start();
	for (size_t i = 0; i < count; i++)
	{
		d.dummy1[i] = 1.0F;
		d.dummy2[i] = 1.0F;
		d.dummy3[i] = 1.0F;
	}
	watch.stop();

	printf("Filling data single loop took: %dms\n", watch.getElapsed());

	watch.start();
	for (size_t i = 0; i < count; i++)
	{
		d.dummy1[i] = 1.0F;
	}
	for (size_t i = 0; i < count; i++)
	{
		d.dummy2[i] = 1.0F;
	}
	for (size_t i = 0; i < count; i++)
	{
		d.dummy3[i] = 1.0F;
	}
	watch.stop();

	printf("Filling data multiple loops took: %dms\n", watch.getElapsed());

	watch.start();
	std::fill_n(d.dummy1, count, 1.0F);
	std::fill_n(d.dummy2, count, 1.0F);
	std::fill_n(d.dummy3, count, 1.0F);
	watch.stop();

	printf("Filling data using std::fill_n took: %dms\n", watch.getElapsed());

	std::cin.get();
	return 0;
}

bigData::bigData(int count)
{
	dummy1 = (float*)malloc(count * sizeof(float));
	dummy2 = (float*)malloc(count * sizeof(float));
	dummy3 = (float*)malloc(count * sizeof(float));
}
```

**Comment by halx99** [comment]
Does the test result is from release build?

**Reply by DelinWorks** [reply]
yes.

**Reply by halx99** [reply]
well, std::fill_n, 1.0f can't be optimized by compiler with memset, just try `0`

**Reply by DelinWorks** [reply]
True it's reduced significantly:
![image](https://user-images.githubusercontent.com/45469625/170135570-42c64728-3d97-4bf3-b90c-1177ca4c14fe.png)

What about separating loops? what's the optimization about it?

**Reply by halx99** [reply]
You can view the asm code by visual studio, it will simply inline `std:fill_n` to invoke system `memset` when fill 0 for float array, it's so fast.
